kv: extract KVBatch service from Internal service and remove api_drpc_hacky

shubhamdhama opened the pull request:
This change removes the old manually-created `api_drpc_hacky.go` file and
uses the properly auto-generated `api_drpc.pb.go` now that the DRPC
compiler setup is working.

We also extract the `KVBatch` service from the `Internal` service, which
has largely been an ad-hoc collection of internal RPCs. This enables:
- Easier Maintenance: Smaller services are simpler to manage and update.
- Smoother DRPC Rollout: We can move to DRPC step-by-step with these
  smaller services, rather than doing `Internal` all at once.
- Subsystems get client interfaces focused only on what they need.

We also refactor `batchStreamImpl` to directly call `RecvMsg` on the
provided stream, which is enabled by the change in codegen to include a
gRPC-compatible `RecvMsg`.

---

Some context on why we use `RecvMsg`: Commit 34e3abfa3 optimized the
`BatchStream` implementation by changing how the `BatchRequest` and its
initial `Requests` slice capacity are allocated. `RecvMsg` allows the
caller to provide a pre-allocated `BatchRequest` instance. This pattern,
when used with a pre-allocated instance of a `BatchRequest` header and
first request together, helps reduce per-message allocations compared to a
`Recv()` method that might allocate a new message object on each call

Fixes: #145179

Release note: none
Epic: CRDB-48934

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145195)
<!-- Reviewable:end -->

shubhamdhama committed:
go.mod: bump drpc to pick codegen changes

This will pickup the following changes:
1. https://github.com/cockroachdb/drpc/pull/3
2. https://github.com/cockroachdb/drpc/pull/2
3. https://github.com/cockroachdb/drpc/pull/4
4. https://github.com/cockroachdb/drpc/pull/6

rickystewart commented on go.mod:
```diff
@@ -505,7 +505,7 @@ replace golang.org/x/time => github.com/cockroachdb/x-time v0.3.1-0.202305251236
 replace github.com/gogo/protobuf => github.com/cockroachdb/gogoproto v1.3.3-0.20241216150617-2358cdb156a1
 
-replace storj.io/drpc => github.com/cockroachdb/drpc v0.0.0-20250507084558-a793c5c40d3d
+replace storj.io/drpc => github.com/shubhamdhama/drpc v0.0.0-20250513152536-21683173691a
```
Hmm, I'm uncertain about us moving from the `cockroachdb/drpc` fork to your personal fork. Is there a reason for this or is this an oversight?

shubhamdhama replied:
Oh, definitely not; we won't be moving out personal fork. I've left a comment here: https://github.com/cockroachdb/cockroach/pull/145195#issuecomment-2888975447 about merging the PRs from the fork first.

I had hoped to check this box before merging, but I realize now I should also add a "stacked on" remark, similar to my other PRs (e.g., https://github.com/cockroachdb/cockroach/pull/146926). I think I should also move this checkbox to the PR description itself.

shubhamdhama replied:
I've updated the PR description, I hope that should make it more explicit. I'm not sure if I should also put a "do-not-merge" as that may imply PR is not ready for review (maybe it's just me).

rickystewart reviewed (commented):
Looks like you have build failures in `pkg/server:server_test`.

rickystewart commented on pkg/kv/kvpb/BUILD.bazel:
```diff
@@ -8,21 +8,17 @@ load(":gen.bzl", "batch_gen")
-        # DRPC protobuf file (api_drpc.pb.go) is currently generated manually.
-        # TODO (chandrat): Remove this once DRPC protobuf generation is
-        # integrated into the build process.
-        "api_drpc_hacky.go",
```
Hmm, I'm not understanding where exactly we added the DRPC stuff to replace this.

shubhamdhama replied:
Right, it's auto-generated and it appears in the tree like this:
```
kvpb
├── ambiguous_result_error.go
├── api_drpc.pb.go
├── api_test.go
├── api.go
├── api.pb.go
```

The PR that auto-generates that is already merged so it must be appearing in everyone's repo.

rickystewart reviewed (approved):
Nice. (As you know, wait until the changes are merged into our fork before merging this one)

shubhamdhama committed:
kv: extract KVBatch service from Internal service and remove api_drpc_hacky

This change removes the old manually-created `api_drpc_hacky.go` file and
uses the properly auto-generated `api_drpc.pb.go` now that the DRPC
compiler setup is working.

We also extract the `KVBatch` service from the `Internal` service, which
has largely been an ad-hoc collection of internal RPCs. This enables:
- Easier Maintenance: Smaller services are simpler to manage and update.
- Smoother DRPC Rollout: We can move to DRPC step-by-step with these
  smaller services, rather than doing `Internal` all at once.
- Subsystems get client interfaces focused only on what they need.

We also refactor `batchStreamImpl` to directly call `RecvMsg` on the
provided stream, which is enabled by the change in codegen to include a
gRPC-compatible `RecvMsg`.

---

Some context on why we use `RecvMsg`: Commit 34e3abfa3 optimized the
`BatchStream` implementation by changing how the `BatchRequest` and its
initial `Requests` slice capacity are allocated. `RecvMsg` allows the
caller to provide a pre-allocated `BatchRequest` instance. This pattern,
when used with a pre-allocated instance of a `BatchRequest` header and
first request together, helps reduce per-message allocations compared to a
`Recv()` method that might allocate a new message object on each call

cthumuluru-crdb commented on DEPS.bzl:
```diff
@@ -11519,10 +11519,10 @@ def go_deps():
-        sha256 = "60246f2907c40615808e4a5af2f6d58a5248802f57a4b7b55dfb03ee5b779b3b",
-        strip_prefix = "github.com/cockroachdb/drpc@v0.0.0-20250507084558-a793c5c40d3d",
+        sha256 = "b68d5d01ee0c45668d1253ba8a31d7e9733047d6c6e7038ce3adf6c1bb43040a",
+        strip_prefix = "github.com/shubhamdhama/drpc@v0.0.0-20250513152536-21683173691a",
```
Please make sure to fix this to use `cockroachdb/drpc` repo before merging the change.

cthumuluru-crdb commented on pkg/server/node.go:
```diff
@@ -1867,18 +1867,10 @@ func (n *Node) Batch(ctx context.Context, args *kvpb.BatchRequest) (*kvpb.BatchR
-	},
-	recvMsg func(*kvpb.BatchRequest) error,
-) error {
+func (n *Node) batchStreamImpl(stream kvpb.RPCKVBatch_BatchStreamStream) error {
```
Do you have changes to generate adapters (ex: `RPCKVBatch_BatchStreamStream`) merged or you are planning to stack this on top of that PR?

shubhamdhama replied:
Yes these also have to be reviewed: [PR #2: codegen: generate common interfaces for stream server and client](https://github.com/cockroachdb/drpc/pull/2) before I can merge current one.

cthumuluru-crdb reviewed (approved):
LGTM. I left a couple of comments. Please make sure to address them before you merge it. Thanks!

shubhamdhama commented:
TFTRs!

bors r=rickystewart,cthumuluru-crdb
